src/firefly/js/ui/SuggestBoxInputField.jsx

```diff
@@ -0,0 +1,175 @@
+import React, {Component, PropTypes}  from 'react';
+import {get, isUndefined} from 'lodash';
+
+import {InputFieldView} from './InputFieldView.jsx';
+import {fieldGroupConnector} from './FieldGroupConnector.jsx';
+import './SuggestBoxInputField.css';
+
+function getProps(params, fireValueChange) {
+    return Object.assign({}, params,
+        {
+            fireValueChange
+        });
+}
+
+const Suggestion = (props) => {
+    const {suggestion, renderSuggestion, ...otherProps} = props;
+    return ( <li {...otherProps}>{renderSuggestion(suggestion)}</li>);
+};
+
+Suggestion.propTypes = {
+    suggestion : PropTypes.any,
+    renderSuggestion: PropTypes.func
+};
+
+const SuggestBox = (props) => {
+    var {suggestions, highlightedIdx, renderSuggestion, onChange, onComplete} = props;
+    suggestions = suggestions || [];
+
+    const handleOptionMouseEnter = (idx, event) => {
+        onChange(idx);
+    };
+
+    const handleOptionClick = (idx, event) => {
+        onComplete(idx);
+    };
+
+    return (
+        <ul className={'SuggestBox'}>
+            {suggestions.map((suggestion, idx) => {
+                return (
+                    <Suggestion
+                        className={'SuggestBox__Suggestion'+(idx === highlightedIdx ? ' SuggestBox__Suggestion--highlighted' : '')}
+                        renderSuggestion={renderSuggestion}
+                        key={idx}
+                        onMouseDown={handleOptionClick.bind(this, idx)} //onClick competes with onBlur
+                        onMouseEnter={handleOptionMouseEnter.bind(this, idx)}
+                        suggestion={suggestion}
+                    />
+                );
+            })}
+        </ul>
+    );
+};
+
+class SuggestBoxInputFieldView extends Component {
+    constructor(props) {
+        super(props);
+        this.state = {
+            isOpen: false,
+            displayValue: props.value,
+            valid: true,
+            message: ''
+        };
+
+        this.onValueChange = this.onValueChange.bind(this);
+    }
+
+
+    onValueChange(ev) {
+        var displayValue = get(ev, 'target.value');
+        var {valid,message} = this.props.validator(displayValue);
+        this.setState({displayValue, valid, message, isOpen: true});
+        this.props.fireValueChange({ value : displayValue, message, valid});
+    }
+
+
+    render() {
+
+        const {displayValue, valid, message, highlightedIdx, isOpen } = this.state;
+        var {validator, label, labelWidth, tooltip, inline, fireValueChange, getSuggestions, renderSuggestion, valueOnSuggestion} = this.props;
+        const currentSuggestions = getSuggestions(displayValue);
+
```

Comment:
I thought `getSuggestions` was going to return a promise? did you decide not to make it async?